### ISSUE TITLE
axis 3 is out of bounds for array of dimension 3( gradcam in py torch/ what is this mean?)

### ISSUE THREAD

**Issue by Zeroroot** [Vooban/Smoothly-Blend-Image-Patches#13]
import torch
import torchvision
import torchvision.transforms as transforms
import torch.nn as nn
import torch.optim as optim
import numpy as np
import cv2
import torch.nn.functional as F

from pytorch_grad_cam import GradCAM
from pytorch_grad_cam.utils.image import show_cam_on_image
from pytorch_grad_cam.utils.model_targets import ClassifierOutputTarget
import timm

def load_and_transform_data():
    transform = transforms.Compose([
        transforms.Resize((224, 224)),
        transforms.ToTensor(),
        transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
    ])

    # CIFAR-10 훈련 데이터셋 로드
    trainset = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)
    trainset = torch.utils.data.Subset(trainset, indices=range(0, len(trainset), 50)) # 데이터셋 크기 축소
    trainloader = torch.utils.data.DataLoader(trainset, batch_size=16, shuffle=True, num_workers=2)

    # CIFAR-10 테스트 데이터셋 로드
    testset = torchvision.datasets.CIFAR10(root='./data', train=False, download=True, transform=transform)
    testset = torch.utils.data.Subset(testset, indices=range(0, len(testset), 1000)) # 데이터셋 크기 축소
    testloader = torch.utils.data.DataLoader(testset, batch_size=32, shuffle=False, num_workers=2)

    return trainloader, testloader

def train_model(model, trainloader):
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=0.003)

    model.train()
    for epoch in range(1):  # 단 한 에포크로 설정
        running_loss = 0.0
        for i, data in enumerate(trainloader, 0):
            inputs, labels = data[0], data[1]

            optimizer.zero_grad()

            outputs = model(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            running_loss += loss.item()
            if i % 100 == 99:
                print(f'[{epoch + 1}, {i + 1}] loss: {running_loss / 100}')
                running_loss = 0.0

    print('Finished Training')

def main():
    device = torch.device("cpu")
    
    trainloader, testloader = load_and_transform_data()
    
    # Loading a ViT model
    model = timm.create_model("vit_small_patch16_224", pretrained=True, num_classes=10)
    model.to(device)

    train_model(model, trainloader)

    model.eval()

    # Grad-CAM 설정: ViT 모델의 마지막 자체 주의 계층을 대상 레이어로 설정
    target_layers = [model.blocks[-1].norm1]

    cam = GradCAM(model=model, target_layers=target_layers)  # Grad-CAM 객체 생성

    dataiter = iter(testloader)
    images, labels = next(dataiter)
    img = images[0].unsqueeze(0)  # 첫 번째 이미지 선택 및 배치 차원 추가

    output = model(img)  # 이미지를 모델에 전달하여 출력 계산
    _, predicted_class = output.max(1)  # 가장 높은 확률을 가진 클래스 선택
    targets = [ClassifierOutputTarget(predicted_class.item())]  # Grad-CAM의 타겟으로 설정할 클래스 지정

    grayscale_cam = cam(input_tensor=img, targets=targets)[0]  # Grad-CAM 생성
    cam_image = np.uint8(grayscale_cam * 255)  # Grayscale CAM 이미지를 0~255 범위로 조정
    
    # 원본 이미지에 CAM 오버레이
    original_image = img.squeeze().numpy().transpose(1, 2, 0)  # 원본 이미지를 numpy 배열로 변환하고 차원 순서 변경
    original_image = (original_image * 0.5 + 0.5) * 255  # 이미지를 [0, 255] 범위로 확장
    original_image = np.uint8(original_image)  # 이미지를 정수형으로 변환

if __name__ == '__main__':
    main()
